click/types.py

```diff
@@ -479,10 +481,51 @@ def convert(self, value, param, ctx):
                     self.path_type,
                     filename_to_ui(value)
                 ), param, ctx)
+            if self.executable and not os.access(value, os.X_OK):
+                self.fail('%s "%s" is not executable.' % (
+                    self.path_type,
+                    filename_to_ui(value)
+                ), param, ctx)
 
         return self.coerce_path_result(rv)
 
 
+class Executable(Path):
+    """The executable type is similar to the :class:`Path` type but checks
+    for an executable file.  A file without path information will be looked
+    up in $PATH.
+
+    For more information see :ref:`tuple-type`.
+
+    :param absolute_path: use the absolute (but not necessarily resolved path).
+    :param resolve_path: if this is true, then the path is fully resolved
+                         before the value is passed onwards.  This means
+                         that it is absolute and symlinks are resolved.  It
+                         will not expand a tilde-prefix, as this is
+                         supposed to be done by the shell only.
+    :param writable: if true, a writable check is performed.
```

Comment:
is this really useful here?

Reply:
well..  I can imagine that you want to look for a readable/writable exe.

Reply:
hrm at least i've never seen any tool that looks up something from $PATH except to execute it - which needs neither of these permissions

Reply:
Yeah.. but I've wanted for this to be generic.. you might want to look for an exe "true" to overwrite then.  Corner case, but it made sense.  The defaults should reflect what you expect from it though.